### ISSUE TITLE
Non-sensical selectors are silently generated

### ISSUE THREAD

**Issue by jacksonrayhamilton** [jsstyles/jss#255]
When I use the API like this:

```js
var jss = require('jss');
var sheet = jss.createStyleSheet({
  '.foo': {
    bar: 0
  }
});
console.log(sheet.toString());
```

...then my expectation is to get:

```css
.foo {
  bar: 0;
}
```

...but instead, I get:

```css
..foo-1028573841 {
  bar: 0;
}
```

Of course, I know that I need to pass `{named: false}` to the `createStyleSheet` to generate the correct class name. Still, non-sensical class name generation can be a "gotcha" for those new to JSS.

The same issue occurs when one forgets to use a plugin, e.g. if `jss-camel-case` is not used, then camel-cased properties will be generated without so much as a warning.

I think it'd be helpful if JSS could detect if it's generating malformed CSS. Maybe this functionality could be available as a development-only plugin, so performance wouldn't need to suffer. It could be as simple as detecting the use of basic CSS syntax / weird casing in keys and printing warnings, or as thorough as a "JSS Lint" utility with knowledge of all valid property names. What do you think?

**Comment by kof** [comment]
Yep, aggree, good idea.

**Comment by kof** [comment]
I think we can even utilize any existing css linter additionally to jss specific linting and lint generated css on the fly. For dev only, of course.

**Comment by jacksonrayhamilton** [comment]
A CSS linter would be a good start, and probably get us error-checking fairly quickly.

However, if we only used a CSS linter to analyze the output, the user would just be told _what_ went wrong with the CSS, rather than precisely which JSON property _caused_ malformed CSS. So it might be good to do checking not just on the output, but also during CSS generation. (Not sure if that's what you meant by "additionally to jss specific linting.")

**Comment by kof** [comment]
Yes the problem might be to have a meaningful explanation where exactly what was wrong if we analyze css output, hopefully we can at least track it down to the rule.

**Comment by kof** [comment]
We are talking about 2 things: one is jss syntax checker for popular errors, second is a css linter.

**Comment by jacksonrayhamilton** [comment]
Yep, they would be two different things. Both seem relatively easy to implement. You described the CSS linter. The "JSS syntax checker" could potentially be embedded inside JSS and its plugins in the form of `if` blocks that throw errors, where those blocks might be stripped out in production by a dead code eliminator. (For an example of this, see [the React source code](https://github.com/facebook/react/blob/d955ee9fae71e2037e9c876e2ab8cb537a8c7e43/src/isomorphic/React.js#L29).)

**Comment by kof** [comment]
What is the benefit  to add it to jss? I want to keep the core small. I would rather see it as  a separate package.

**Comment by jacksonrayhamilton** [comment]
Looks like that checking could work as a package, since the plugin API provides access to `rule.style`, which we could examine.

A question that lingers is, where do plugin-specific checks go?
- In the case of the jss-nested issue I mentioned, the "&" check could potentially (and appropriately) be performed in the jss-nested plugin.
- But in the case of detecting the absence of the jss-camel-case plugin, that would logically have to be done external to that plugin.

So all the checks for all the plugins could potentially be in one "jss-lint" plugin. Which has the potential to be messy, but then again it might be just fine, especially if we limited the linting to the just plugins maintained by the jsstyles organization.

**Comment by kof** [comment]
Yeah, it would make sense to maintain plug related checks within plugin packages. So that once plugin changes, it immediately changes int's linter.

**Comment by kof** [comment]
How about this: 

1. one linter package
1. it lints the stuff from the core
1. every jss plugin can also implement an optional linter plugin, basically plugins can plug-in to jss as well as into linter. It can check the absence of itself in the setup as well as malformed rules
1. linter uses all plugins packages as dependencies

Example of such a plugin with interface:

```javascript

export default function jssExtend() {
  return (rule) => {...}
}

export function jssExtendLinter() {
  return (rule) => {...}
}
```

**Comment by jacksonrayhamilton** [comment]
Seems pretty good. Maybe the linting code should be in a separate file, though? (Stored in `lint.js`, alongside the plugin's `index.js`.) That way, the linting code wouldn't affect the size of the production bundle.

**Comment by jacksonrayhamilton** [comment]
As for the CSS linter used on the generated CSS output: Should we pick one, or allow that to be customized, too?

It might be easier to set up just one CSS linter. Although, it might be harder for users to configure it, and also the choice would be opinionated. (I've actually only used SCSS linters, so I'm not in a position to judge.)

If the CSS linter were to be customizable, the user would have maximum freedom in that regard, but we'd have to define a compatibility interface (probably a function that takes a CSS string as input and reports errors in a format that jss-lint understands). Also, some linters like stylelint return a Promise, so the API should probably support that, too.

**Comment by kof** [comment]
https://github.com/stylelint/stylelint

**Comment by kof** [comment]
- [ ] warn about plugins used in the wrong order

**Comment by bytasv** [comment]
Hey guys, does linting currently work in any way?

**Comment by kof** [comment]
Nope, so far nobody had time to work on it.
